### PR TITLE
AND #158 Implement pending request logic.

### LINKED ISSUE #158: Implement Pending Requests page

### PR DESCRIPTION
**Pull request by JuliusAlexanderIV**
closes #158

### CODE REVIEW

File: app/src/main/java/xyz/whereuat/whereuat/Constants.java [hunk 1]
```diff
@@ -25,10 +25,15 @@
...
 
+    public static final String CONTACT_NUMBER_BUNDLE_EXTRA = "CONTACT_NUMBER_BUNDLE_EXTRA";
+
     // Constants related to the broadcasts sent by the application.
     public static final String AT_RESPONSE_INITIATE_BROADCAST =
             "xyz.whereuat.whereuat.AT_RESPONSE_INITIATE_BROADCAST";
     public static final String TOKEN_BROADCAST = "xyz.whereuat.whereuat.TOKEN_BROADCAST";
+    public static final String RELOAD_PENDING_REQS_BROADCAST =
+            "xyz.whereuat.RELOAD_PENDING_REQS_BROADCAST";
+    public static final String RELOAD_CONTACTS_BROADCAST = "whereuat.xyz.RELOAD_CONTACTS_BROADCAST";
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/Constants.java]: xyz.whereuat.whereuat.RELOAD_CONTACTS_BROADCAST

File: app/src/main/java/xyz/whereuat/whereuat/Constants.java [hunk 2]
```diff
@@ -25,10 +25,15 @@
...
     public static final String TO_PHONE_EXTRA = "TO_PHONE_EXTRA";
 
+    public static final String CONTACT_NUMBER_BUNDLE_EXTRA = "CONTACT_NUMBER_BUNDLE_EXTRA";
+
     // Constants related to the broadcasts sent by the application.
     public static final String AT_RESPONSE_INITIATE_BROADCAST =
             "xyz.whereuat.whereuat.AT_RESPONSE_INITIATE_BROADCAST";
     public static final String TOKEN_BROADCAST = "xyz.whereuat.whereuat.TOKEN_BROADCAST";
+    public static final String RELOAD_PENDING_REQS_BROADCAST =
+            "xyz.whereuat.RELOAD_PENDING_REQS_BROADCAST";
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/Constants.java]: xyz.whereuat.whereuat.RELOAD_PENDING_REQS_BROADCAST

File: app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java [hunk 3]
```diff
@@ -73,26 +76,72 @@ public void run() {
...
+                    WuaGcmListenerService.this, name, from_phone).call() != -1;
+
+            NotificationUtils.sendPendingRequestNotification(WuaGcmListenerService.this,
+                    from_phone);
+
+            if (was_inserted) {
+                PendingRequestsActivity.notifyOfPendingRequestChange(WuaGcmListenerService.this);
+                Log.d(TAG, "Contact was added to pending requests.");
+            } else {
+                Log.d(TAG, "Contact was unsuccessfully added to pending requests.");
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java]: huh?

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    @Override
+    public Dialog onCreateDialog(final Bundle saved_instance_state) {
+        final Activity activity = getActivity();
+
+        AlertDialog.Builder builder = new AlertDialog.Builder(activity, R.style.DialogTheme);
+        LayoutInflater inflater = activity.getLayoutInflater();
+        View view = inflater.inflate(R.layout.add_contact_dialog, (ViewGroup) this.getView());
+        builder.setView(view)
+                .setTitle(R.string.add_contact_prompt_title_text)
+                // When clicked, the "Okay" button should add the key location to the database if it
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: Nice copy-paste antipattern bro

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                // is valid.
+                .setPositiveButton(R.string.okay, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {
+                        Dialog d = (Dialog) dialog;
+                        EditText e = (EditText) d.findViewById(R.id.contact_name_input);
+
+                        String name = e.getText().toString();
+
+                        // Make sure the name is not null or empty and the location is not null.
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: Comment talks about location

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                if (hasFocus) {
+                    d.getWindow().setSoftInputMode(
+                            WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
+                }
+            }
+        });
+    }
+
+    /**
+     * This function wraps an insert into the pending requests table.
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: Isn't this inserting into the Contacts table?

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+     *
+     * @param activity the Activity that the fragment comes from
+     * @param name the name of the new contact
+     * @param number the number of the new contact
+     */
+    private void addContact(final Activity activity, final String name, final String number) {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: Can we rename this to make it clear it's also deleting from the pending requests table? Maybe `addContactFromPending` or something

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                // The "Cancel" button shouldn't do anything.
+                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(DialogInterface dialog, int which) {}
+                });
+        return builder.create();
+    }
+
+    @Override
+    public void onStart() {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: Can we move this method? It's kind of just hanging out awkwardly between two closely-linked methods

File: app/src/main/java/xyz/whereuat/whereuat/utils/ContactUtils.java [hunk 9]
```diff
@@ -80,6 +83,17 @@ public static InsertCommand buildInsertCommand(Context context, String name, Str
         return new InsertCommand(context, ContactEntry.TABLE_NAME, null, values);
     }
 
+    public static int generateRandomColor() {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/utils/ContactUtils.java]: I wouldn't be against moving this into its own utility either. But I also have no issue with leaving that for a later PR when we actually figure out what we're doing with colors.

File: app/src/main/java/xyz/whereuat/whereuat/utils/PendingRequestUtils.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    public static InsertCommand buildInsertCommand(Context context, String name, String phone) {
+        ContentValues values = new ContentValues();
+        values.put(PendingRequestEntry.COLUMN_NAME, name);
+        values.put(PendingRequestEntry.COLUMN_PHONE, phone);
+        return new InsertCommand(context, PendingRequestEntry.TABLE_NAME, null, values);
+    }
+
+    public static QueryCommand buildSelectAllCommand(Context context) {
+        return new QueryCommand(context, PendingRequestEntry.TABLE_NAME, false,
+                PendingRequestEntry.COLUMNS, null, null, null, null, null, null);
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/utils/PendingRequestUtils.java]: I completely forgot about this for the previous PR, but an argument of `null` automatically returns all columns -- no need for the `.COLUMNS` member here

File: app/src/main/java/xyz/whereuat/whereuat/utils/PhonebookUtils.java [hunk 11]
```diff
@@ -47,6 +50,33 @@
...
+     */
+    @NonNull
+    public static String queryPhonebookForContactName(Context context, String phone) {
+        Uri uri = Uri.withAppendedPath(ContactsContract.PhoneLookup.CONTENT_FILTER_URI, phone);
+        Cursor cursor = context.getContentResolver().query(uri,
+                new String[] {ContactsContract.PhoneLookup.DISPLAY_NAME}, null, null, null);
+
+        String name = "";
+
+        if (cursor.moveToFirst()) {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/utils/PhonebookUtils.java]: Is `cursor` guaranteed to be non-null? Could be cleaner to use a try-catch-finally here to handle closing the cursor too

File: app/src/main/res/layout/pending_req_list_item.xml [hunk 12]
```diff
@@ -56,4 +56,4 @@
         android:scaleType="fitCenter"
         android:src="@drawable/ic_delete"
         android:tint="@color/darkGray"/>
-</LinearLayout>
\ No newline at end of file
+</LinearLayout>
```
Comment (kangp3) [inline app/src/main/res/layout/pending_req_list_item.xml]: Nice.

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+import xyz.whereuat.whereuat.PendingRequestsActivity;
+import xyz.whereuat.whereuat.R;
+import xyz.whereuat.whereuat.utils.ContactUtils;
+import xyz.whereuat.whereuat.utils.PendingRequestUtils;
+
+/**
+ * This class is used for handling the dialog that is displayed when the user tries to add pending
+ * request for a number that was not found in the phonebook.
+ */
+public class AddContactDialogFragment extends DialogFragment {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: I'd prefer to change name of this to `AddPendingContactDialogFragment` to make it clear that this is not for adding a generic contact

Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: However, if we do decide to use this to allow users to add contacts without granting us phonebook permissions I'm very okay with the name.

File: app/src/main/java/xyz/whereuat/whereuat/MainActivity.java [hunk 14]
```diff
@@ -301,15 +314,27 @@ public void run() {
...
-        Random rnd = new Random();
-        // A number between 0 and 360.
-        float hue = rnd.nextInt(360);
-        // The multiplier keeps the value in a range, the addition keeps the number farther from 0
-        // so colors that are almost black aren't generated.
-        float value = rnd.nextFloat() * 0.4f + 0.4f;
-        // Keep the saturation constant at 0.3.
-        return Color.HSVToColor(new float[] {hue, 0.3f, value});
+    /**
+     * This funciton is so other classes can notify this activity that the contacts have changed
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/MainActivity.java]: s/funciton/function

File: app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java [hunk 15]
```diff
@@ -67,11 +79,32 @@ public void onCreate(Bundle saved_instance_state) {
...
+    public void onLoadFinished(Loader<Cursor> loader, Cursor data) { mAdapter.swapCursor(data); }
 
     /**
      * This is called when the last Cursor provided to onLoadFinished() is about to be closed.
      * Swapping with null ensures the cursor is no longer being used.
      */
     public void onLoaderReset(Loader<Cursor> loader) { mAdapter.swapCursor(null); }
+
+    /**
+     * This funciton is so other classes can notify this activity that the pending requests have
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java]: s/funciton/function

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java [hunk 16]
```diff
@@ -40,13 +56,91 @@ public void bindView (View view, Context context, Cursor cursor) {
...
             public void onClick(View v) {
-                Toast.makeText(mContext, "Adding contact...", Toast.LENGTH_SHORT).show();
+                if (name.isEmpty()) {
+                    handleWithAddContactDialog(number);
+                } else {
+                    addContact(context, name, number);
+                }
             }
         });
         (view.findViewById(R.id.delete_btn)).setOnClickListener(new View.OnClickListener() {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: Not part of the commit, but this doesn't need to be wrapped in parens

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java [hunk 17]
```diff
@@ -40,13 +56,91 @@ public void bindView (View view, Context context, Cursor cursor) {
         (view.findViewById(R.id.add_btn)).setOnClickListener(new View.OnClickListener() {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: Not part of the commit, but this doesn't need to be wrapped in parens

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java [hunk 18]
```diff
@@ -40,13 +56,91 @@ public void bindView (View view, Context context, Cursor cursor) {
...
+    }
+
+    /**
+     * This function handles adding the contact to the database.
+     *
+     * @param context the context to execute the database insert from
+     * @param name the new contact's name
+     * @param number the new contact's number
+     */
+    private void addContact(final Context context, final String name, final String number) {
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: Would it be better to have `AddContactDialogFragment` be a subclass of `PendingRequestAdapter` to prevent this double-declaration stuff? Could use a similar structure for the `KeyLocDialogFragment`. In any case, out of scope of this commit

File: app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java [hunk 19]
```diff
@@ -1,48 +1,55 @@
...
-    PendingRequestAdapter mAdapter;
-    MatrixCursor mCursor;
+    private PendingRequestAdapter mAdapter;
+    private ReloadPendingRequestsReceiver mReloadPendingReqsReceiver;
+    private IntentFilter mReloadPendingReqsFilter;
 
     @Override
     public void onCreate(Bundle saved_instance_state) {
         super.onCreate(saved_instance_state);
         setContentView(R.layout.activity_pending_reqs);
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java]: Toolbar for the pending requests activity should say "Pending Requests" instead of "whereu@"

**Comment by kangp3** [comment]
Done with first round of comments.

File: app/src/main/java/xyz/whereuat/whereuat/Constants.java (hunk 2 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/Constants.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/Constants.java (hunk 1 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/Constants.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/MainActivity.java (hunk 14 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/MainActivity.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java (hunk 15 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/PendingRequestsActivity.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java (hunk 3 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java]: what specifically are you "huh"ing about? haha

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java (hunk 16 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: added to stupids

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java (hunk 17 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: added to stupids

File: app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java (hunk 18 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/adapters/PendingRequestAdapter.java]: we talked about this, it sucks. we'll figure out a solution

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 13 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: A little confused by what you mean here, however I do like the term "pending contact". I keep finding "pending request" to be rather confusing because the requests don't actually pend, they just add the contact to holding area until you approve them.

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 4 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: fixed

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 5 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: fixed

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 8 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 6 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: yes, fixed

File: app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java (hunk 7 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/ui/views/AddContactDialogFragment.java]: `addContactFromPending`

File: app/src/main/java/xyz/whereuat/whereuat/utils/ContactUtils.java (hunk 9 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/utils/ContactUtils.java]: we'll figure out later then

File: app/src/main/java/xyz/whereuat/whereuat/utils/PendingRequestUtils.java (hunk 10 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/utils/PendingRequestUtils.java]: done

File: app/src/main/java/xyz/whereuat/whereuat/utils/PhonebookUtils.java (hunk 11 above)
Comment (JuliusAlexanderIV) [inline app/src/main/java/xyz/whereuat/whereuat/utils/PhonebookUtils.java]: This guy says so and the docs say `query` returns a `Cursor` object.
http://stackoverflow.com/questions/12445010/what-is-the-use-of-movetofirst-in-sqlite-cursors

File: app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java [hunk 20]
```diff
@@ -119,25 +119,25 @@ private void handleKnownContact(Cursor contact) {
...
                     from_phone);
 
             if (was_inserted) {
-                PendingRequestsActivity.notifyOfPendingRequestChange(WuaGcmListenerService.this);
-                Log.d(TAG, "Contact was added to pending requests.");
+                ContactRequestsActivity.notifyOfContactRequestChange(WuaGcmListenerService.this);
+                Log.d(TAG, "Contact was added to contact requests.");
             } else {
-                Log.d(TAG, "Contact was unsuccessfully added to pending requests.");
+                Log.d(TAG, "Contact was unsuccessfully added to contact requests.");
```
Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java]: Wording fix

Comment (kangp3) [inline app/src/main/java/xyz/whereuat/whereuat/gcm/WuaGcmListenerService.java]: Wording fix

**Comment by kangp3** [comment]
lgtm after one comment
